File: dynamicProgrammingSamples/canSum.js

const canSum = (target, collection) => {
    if (target == 0) return true;
    if (target < 0) return false;

    for(let num of collection){
        const reminder = target - num;
        if (canSum(reminder, collection) == true) return true;
    }
    return false;
}

const canSumMem = (target, collection, memo = {}) => {
    if(target in memo) return memo[target];
    if (target == 0) return true;
    if (target < 0) return false;

    for(let num of collection){
        const reminder = target - num;
        memo[target] = canSumMem(reminder, collection, memo);
        if (memo[target] == true) return true;
    }
    memo[target] = false;
    return false;
}

const canSumTab = (target, collection) => {
    const tab = Array(target + 1).fill(false);
    tab[0] = true;
    for(let i = 0; i <= target; i++){
        if(!tab[i]) continue;
        for(let num of collection){
            tab[i + num] = true;
        }
    }
    return tab[target];
}

/* console.log(canSum(5, [1, 2, 3, 4, 5])) // true
console.log(canSum(7, [2, 4])) // false
console.log(canSum(8, [2, 3, 5])) // true
console.log(canSum(300, [7, 14])) // false */

/* console.log(canSumMem(5, [1, 2, 3, 4, 5])) // true
console.log(canSumMem(7, [2, 4])) // false
console.log(canSumMem(8, [2, 3, 5])) // true
console.log(canSumMem(300, [7, 14])) // false  */

console.log(canSumTab(5, [1, 2, 3, 4, 5])) // true
console.log(canSumTab(7, [2, 4])) // false
console.log(canSumTab(8, [3, 5])) // true
console.log(canSumTab(300, [7, 14])) // false 

File: dynamicProgrammingSamples/cpp-version/canSum.cpp
#include <iostream>
#include <vector>

using namespace std;

bool canSum(int target, vector<int> collection);
bool canSumTab(int target, vector<int> collection);


int main(){
    vector<int> tab1{1, 2, 3, 4, 5};
    vector<int> tab2{2, 4};
    cout << "Recursion method result on (5,[1, 2, 3, 4, 5]) && (7,[2, 4])" << endl;
    cout << canSum(5, tab1) << endl; // true
    cout << canSum(7, tab2) << endl; // false
    cout << endl;
    cout << "Tabulation method result on (5,[1, 2, 3, 4, 5]) && (7,[2, 4])" << endl;
    cout << canSum(5, tab1) << endl; // true
    cout << canSum(7, tab2) << endl; // false
    return 0;
}

bool canSum(int target, vector<int> collection){
    if (target == 0) return true;
    if (target < 0) return false;

    for(int i = 0; i < collection.size(); i++){
        int reminder = target - collection.at(i);
        if (canSum(reminder, collection)) return true;
    }
    return false;
}

bool canSumTab(int target, vector<int> collection){
    vector<bool> tab;
    tab.assign(target + 1, false);
    tab[0] = true;
    for(int i=0; i <= target; i++){
        if(!tab.at(i)) continue;
        for(int num=0; num < collection.size(); i++){
            tab[i + num] = true;
        }
    }
    return tab.at(target);
}

